organizations/data.py

```diff
@@ -217,6 +279,87 @@ def create_organization_course(organization, course_key):
         )
 
 
+def bulk_create_organization_courses(organization_course_pairs):
+    """
+    Efficiently insert multiple organization-course relationships into the database.
+
+    Arguments:
+
+        organization_course_pairs (iterable[tuple[dict, CourseKey]]):
+
+            An iterable of (organization_data, course_key) pairs.
+
+            Organization-course linkages that DO NOT already exist will be created.
+            Organization-course linkages that DO already exist will be activated, if inactive.
+
+            Assumption: All provided organizations already exist in the DB.
+    """
+    # For sake of having sane variable names, please understand
+    #   * "orgslug" to mean "lowercase organization short name" and
+    #   * "courseid" to mean "stringified course run key"
+    # in the context of this function.
+    # We normalize short_names to lowercase because MySQL UNIQUE is case-insensitive.
+
+    # Build set of pairs: (lowercased org short name, course key string).
+    orgslug_courseid_pairs = {
+        (organization_data["short_name"].lower(), str(course_key))
+        for organization_data, course_key
+        in organization_course_pairs
+    }
+
+    # Grab all existing (lowercased org short name, course key string) pairs from db,
+    # filtering for the ones requested for creation in `organization_course_pairs`, and
+    # indexing by db `id` of org_course object (we need this later for the bulk update).
+    orgslug_courseid_pairs_to_activate_by_id = {
+        org_course_id: (short_name.lower(), course_id)
+        for org_course_id, short_name, course_id
+        in internal.OrganizationCourse.objects.values_list(
+            "id", "organization__short_name", "course_id"
+        )
+        if (short_name.lower(), course_id) in orgslug_courseid_pairs
+    }
+
+    # Working backwards from the set of pairs we need to *activate*,
+    # find the set of pairs we need to *create*.
```

Comment:
I really appreciate the commenting btw. There's a lot of dict creation that could be confusing, but you walk through it well.

Reply:
Glad you found it helpful!